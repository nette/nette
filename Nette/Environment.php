<?php

/**
 * Nette Framework
 *
 * Copyright (c) 2004, 2009 David Grudl (http://davidgrudl.com)
 *
 * This source file is subject to the "Nette license" that is bundled
 * with this package in the file license.txt.
 *
 * For more information please see http://nettephp.com
 *
 * @copyright  Copyright (c) 2004, 2009 David Grudl
 * @license    http://nettephp.com/license  Nette license
 * @link       http://nettephp.com
 * @category   Nette
 * @package    Nette
 * @version    $Id$
 */

/*namespace Nette;*/



/**
 * Nette environment and configuration.
 *
 * @author     David Grudl
 * @copyright  Copyright (c) 2004, 2009 David Grudl
 * @package    Nette
 */
final class Environment
{
	/**#@+ environment name */
	const DEVELOPMENT = 'development';
	const PRODUCTION = 'production';
	const CONSOLE = 'console';
	const LAB = 'lab';
	/**#@-*/

	/**#@+ mode name */
	const DEBUG = 'debug';
	const PERFORMANCE = 'performance';
	/**#@-*/

	/** @var Configurator */
	private static $configurator;

	/** @var string  the mode of current application */
	private static $mode = array();

	/** @var \ArrayObject */
	private static $config;

	/** @var IServiceLocator */
	private static $serviceLocator;

	/** @var array */
	private static $vars = array(
		'encoding' => array('UTF-8', FALSE),
		'lang' => array('en', FALSE),
		'cacheBase' => array('%tempDir%/cache-', TRUE),
		'tempDir' => array('%appDir%/temp', TRUE),
		'logDir' => array('%appDir%/log', TRUE),
		'templatesDir' => array('%appDir%/templates', TRUE),
		'presentersDir' => array('%appDir%/presenters', TRUE),
		'componentsDir' => array('%appDir%/components', TRUE),
		'modelsDir' => array('%appDir%/models', TRUE),
	);



	/**
	 * Static class - cannot be instantiated.
	 */
	final public function __construct()
	{
		throw new /*\*/LogicException("Cannot instantiate static class " . get_class($this));
	}



	/**
	 * Sets "class behind Environment" configurator.
	 * @param  Configurator
	 * @return void
	 */
	public static function setConfigurator(Configurator $configurator)
	{
		self::$configurator = $configurator;
	}



	/**
	 * Gets "class behind Environment" configurator.
	 * @return Configurator
	 */
	public static function getConfigurator()
	{
		if (self::$configurator === NULL) {
			self::$configurator = new Configurator;
		}
		return self::$configurator;
	}



	/********************* environment name and modes ****************d*g**/



	/**
	 * Sets the current environment name.
	 * @param  string
	 * @return void
	 * @throws \InvalidStateException
	 */
	public static function setName($name)
	{
		if (!isset(self::$vars['environment'])) {
			self::setVariable('environment', $name, FALSE);

		} else {
			throw new /*\*/InvalidStateException('Environment name has been already set.');
		}
	}



	/**
	 * Returns the current environment name.
	 * @return string
	 */
	public static function getName()
	{
		$name = self::getVariable('environment');
		if ($name === NULL) {
			$name = self::getConfigurator()->detect('environment');
			self::setVariable('environment', $name, FALSE);
		}
		return $name;
	}



	/**
	 * Sets the mode.
	 *
	 * @param  string mode identifier
	 * @param  bool   set or unser
	 * @return void
	 */
	public static function setMode($mode, $flag = TRUE)
	{
		if ($mode === 'live') {
			trigger_error("Environment mode 'live' is deprecated; use 'production' instead.", E_USER_WARNING);
			$mode = 'production';
		}
		self::$mode[$mode] = (bool) $flag;
	}



	/**
	 * Returns the mode.
	 *
	 * @param  string mode identifier
	 * @return bool
	 */
	public static function getMode($mode)
	{
		if ($mode === 'live') {
			trigger_error("Environment mode 'live' is deprecated; use 'production' instead.", E_USER_WARNING);
			$mode = 'production';
		}
		if (isset(self::$mode[$mode])) {
			return self::$mode[$mode];

		} else {
			return self::$mode[$mode] = self::getConfigurator()->detect($mode);
		}
	}



	/**
	 * Detects console (non-HTTP) mode.
	 * @return bool
	 */
	public static function isConsole()
	{
		return self::getMode('console');
	}



	/**
	 * Determines whether a server is running in production mode.
	 * @return bool
	 */
	public static function isProduction()
	{
		return self::getMode('production');
	}



	/**
	 * @deprecated {@link Environment::isProduction()}
	 */
	public static function isLive()
	{
		trigger_error('Environment::isLive() is deprecated; use Environment::isProduction() instead.', /**/E_USER_WARNING/**//*E_USER_DEPRECATED*/);
		return self::getMode('production');
	}



	/**
	 * Determines whether the debugger is active.
	 * @return bool
	 */
	public static function isDebugging()
	{
		return self::getMode('debug');
	}



	/********************* environment variables ****************d*g**/



	/**
	 * Sets the environment variable.
	 * @param  string
	 * @param  mixed
	 * @param  bool
	 * @return void
	 */
	public static function setVariable($name, $value, $expand = TRUE)
	{
		if (!is_string($value)) {
			$expand = FALSE;
		}
		self::$vars[$name] = array($value, (bool) $expand);
	}



	/**
	 * Returns the value of an environment variable or $default if there is no element set.
	 * @param  string
	 * @param  mixed  default value to use if key not found
	 * @return mixed
	 * @throws \InvalidStateException
	 */
	public static function getVariable($name, $default = NULL)
	{
		if (isset(self::$vars[$name])) {
			list($var, $exp) = self::$vars[$name];
			if ($exp) {
				$var = self::expand($var);
				self::$vars[$name] = array($var, FALSE);
			}
			return $var;

		} else {
			// convert from camelCaps (or PascalCaps) to ALL_CAPS
			$const = strtoupper(preg_replace('#(.)([A-Z]+)#', '$1_$2', $name));
			$list = get_defined_constants(TRUE);
			if (isset($list['user'][$const])) {
				self::$vars[$name] = array($list['user'][$const], FALSE);
				return $list['user'][$const];

			} else {
				return $default;
			}
		}
	}



	/**
	 * Define one or more variables as constants.
	 * @param  string|array
	 * @return void
	 */
	public static function exportConstant($names)
	{
		if (!is_array($names)) {
			$names = func_get_args();
		}

		foreach ($names as $name) {
			$const = strtoupper(preg_replace('#(.)([A-Z]+)#', '$1_$2', $name));
			define($const, self::getVariable($name));
		}
	}



	/**
	 * Returns expanded variable.
	 * @param  string
	 * @return string
	 * @throws \InvalidStateException
	 */
	public static function expand($var)
	{
		if (is_string($var) && strpos($var, '%') !== FALSE) {
			return @preg_replace_callback('#%([a-z0-9_-]*)%#i', array(__CLASS__, 'expandCb'), $var); // intentionally @ due PHP bug #39257
		}
		return $var;
	}



	/**
	 * @see Environment::expand()
	 * @param  array
	 * @return string
	 */
	private static function expandCb($m)
	{
		list(, $var) = $m;
		if ($var === '') return '%';

		static $livelock;
		if (isset($livelock[$var])) {
			throw new /*\*/InvalidStateException("Circular reference detected for variables: "
				. implode(', ', array_keys($livelock)) . ".");
		}

		try {
			$livelock[$var] = TRUE;
			$val = self::getVariable($var);
			unset($livelock[$var]);
		} catch (/*\*/Exception $e) {
			$livelock = array();
			throw $e;
		}

		if ($val === NULL) {
			throw new /*\*/InvalidStateException("Unknown environment variable '$var'.");

		} elseif (!is_scalar($val)) {
			throw new /*\*/InvalidStateException("Environment variable '$var' is not scalar.");
		}

		return $val;
	}



	/********************* service locator ****************d*g**/



	/**
	 * Get initial instance of service locator.
	 * @return IServiceLocator
	 */
	public static function getServiceLocator()
	{
		if (self::$serviceLocator === NULL) {
			self::$serviceLocator = self::getConfigurator()->createServiceLocator();
		}
		return self::$serviceLocator;
	}



	/**
	 * Gets the service object of the specified type.
	 * @param  string service name
	 * @param  bool   throw exception if service doesn't exist?
	 * @return mixed
	 */
	static public function getService($name, $need = TRUE)
	{
		return self::getServiceLocator()->getService($name, $need);
	}



	/**
	 * @return Nette\Web\IHttpRequest
	 */
	public static function getHttpRequest()
	{
		return self::getServiceLocator()->getService('Nette\Web\IHttpRequest');
	}



	/**
	 * @return Nette\Web\IHttpResponse
	 */
	public static function getHttpResponse()
	{
		return self::getServiceLocator()->getService('Nette\Web\IHttpResponse');
	}



	/**
	 * @return Nette\Application\Application
	 */
	public static function getApplication()
	{
		return self::getServiceLocator()->getService('Nette\Application\Application');
	}



	/**
	 * @return Nette\Web\IUser
	 */
	public static function getUser()
	{
		return self::getServiceLocator()->getService('Nette\Web\IUser');
	}



	/********************* service factories ****************d*g**/



	/**
	 * @param  string
	 * @return Nette\Caching\Cache
	 */
	public static function getCache($namespace = '')
	{
		return new /*Nette\Caching\*/Cache(
			self::getService('Nette\Caching\ICacheStorage'),
			$namespace
		);
	}



	/**
	 * Returns instance of session or session namespace.
	 * @param  string
	 * @return Nette\Web\Session|Nette\Web\Session
	 */
	public static function getSession($namespace = NULL)
	{
		$handler = self::getService('Nette\Web\Session');
		return func_num_args() === 0 ? $handler : $handler->getNamespace($namespace);
	}



	/********************* global configuration ****************d*g**/



	/**
	 * Loads global configuration from file and process it.
	 * @param  string|Nette\Config\Config  file name or Config object
	 * @return \ArrayObject
	 */
	public static function loadConfig($file = NULL)
	{
		return self::$config = self::getConfigurator()->loadConfig($file);
	}



	/**
	 * Returns the global configuration.
	 * @param  string key
	 * @param  mixed  default value
	 * @return mixed
	 */
	public static function getConfig($key = NULL, $default = NULL)
	{
		if (func_num_args()) {
			return isset(self::$config[$key]) ? self::$config[$key] : $default;

		} else {
			return self::$config;
		}
	}

}
