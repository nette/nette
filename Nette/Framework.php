<?php

/**
 * This file is part of the Nette Framework (https://nette.org)
 * Copyright (c) 2004 David Grudl (https://davidgrudl.com)
 */

declare(strict_types=1);

namespace Nette;


/**
 * Framework version.
 */
class Framework
{
	public const
		NAME = 'Nette Framework',
		VERSION = '3.0',
		VERSION_ID = 30000,
		REVISION = '';
}
