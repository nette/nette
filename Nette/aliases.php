<?php

defined('NETTE_ALIASES') && empty(NETTE_ALIASES) || array_walk(new ArrayIterator(array(
	'Nette\Config\Configurator' => 'Nette\Configurator',
	'Nette\Config\CompilerExtension' => 'Nette\DI\CompilerExtension',
	'Nette\Diagnostics\Bar' => 'Tracy\Bar',
	'Nette\Diagnostics\BlueScreen' => 'Tracy\BlueScreen',
	'Nette\Diagnostics\Dumper' => 'Tracy\Dumper',
	'Nette\Diagnostics\FireLogger' => 'Tracy\FireLogger',
	'Nette\Diagnostics\Logger' => 'Tracy\Logger',
	'Nette\Latte\ParseException' => 'Latte\CompileException',
	'Nette\Latte\CompileException' => 'Latte\CompileException',
	'Nette\Latte\Compiler' => 'Latte\Compiler',
	'Nette\Latte\IMacro' => 'Latte\IMacro',
	'Nette\Latte\Parser' => 'Latte\Parser',
	'Nette\Latte\Macros\CoreMacros' => 'Latte\Macros\CoreMacros',
	'Nette\Latte\Macros\MacroSet' => 'Latte\Macros\MacroSet',
	'Nette\ArrayHash' => 'Nette\Utils\ArrayHash',
	'Nette\ArrayList' => 'Nette\Utils\ArrayList',
	'Nette\DateTime' => 'Nette\Utils\DateTime',
	'Nette\Image' => 'Nette\Utils\Image',
	'Nette\ObjectMixin' => 'Nette\Utils\ObjectMixin',
)), 'class_alias');
