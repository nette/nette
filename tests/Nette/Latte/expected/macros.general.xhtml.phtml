<?php
// prolog Nette\Latte\Macros\CoreMacros
list($_l, $_g) = Nette\Latte\Macros\CoreMacros::initRuntime($template, 'xxx')
;
// prolog Nette\Latte\Macros\UIMacros
//
// block menu
//
if (!function_exists($_l->blocks['menu'][] = '_xxx_menu')) { function _xxx_menu($_l, $_args) { extract($_args)
?><ul>
<?php $iterations = 0; foreach ($iterator = $_l->its[] = new Nette\Iterators\CachingIterator($menu) as $item): ?>
	<li><?php echo Nette\Templating\Helpers::escapeHtml($counter++, ENT_NOQUOTES) ?>
 <?php if (is_array($item)): ?> <?php call_user_func(reset($_l->blocks['menu']), $_l, array('menu' => $item) + get_defined_vars()) ?>
 <?php else: echo Nette\Templating\Helpers::escapeHtml($item, ENT_NOQUOTES) ;endif ?></li>
<?php $iterations++; endforeach; array_pop($_l->its); $iterator = end($_l->its) ?>
</ul>
<?php
}}

//
// block bl
//
if (!function_exists($_l->blocks['bl'][] = '_xxx_bl')) { function _xxx_bl($_l, $_args) { extract($_args)
?><ul title="block + if + foreach">
<?php $iterations = 0; foreach ($people as $person): if (strlen($person)===4): ?>
	<li><?php echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ?></li>
<?php endif ;$iterations++; endforeach ?>
</ul>
<?php
}}

//
// end of blocks
//

// template extending and snippets support

$_l->extends = empty($template->_extended) && isset($_control) && $_control instanceof Nette\Application\UI\Presenter ? $_control->findLayoutTemplateFile() : NULL; $template->_extended = $_extended = TRUE;


if ($_l->extends) {
	%A%}

//
// main template
//
?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<ul>
	<li>Escaped: <?php echo Nette\Templating\Helpers::escapeHtml($hello, ENT_NOQUOTES) ?></li>
	<li>Non-escaped: <?php echo $hello ?></li>
	<li>Escaped expression: <?php echo Nette\Templating\Helpers::escapeHtml('<' . 'b' . '>hello' . '</b>', ENT_NOQUOTES) ?></li>
	<li>Non-escaped expression: <?php echo '<' . 'b' . '>hello' . '</b>' ?></li>
	<li>Array access: <?php echo Nette\Templating\Helpers::escapeHtml($people[1], ENT_NOQUOTES) ?></li>
	<li>Html: <?php echo Nette\Templating\Helpers::escapeHtml($el, ENT_NOQUOTES) ?></li>
	<li>Translated: <?php echo Nette\Templating\Helpers::escapeHtml($template->translate($people[1]), ENT_NOQUOTES) ?></li>
	<li>Translated string: <?php echo Nette\Templating\Helpers::escapeHtml($template->translate('ahoj'), ENT_NOQUOTES) ?></li>
	<li>Non-escaped and translated: <?php echo $template->translate($people[1]) ?></li>
	<li>Condition: <?php echo Nette\Templating\Helpers::escapeHtml($hello ? 'yes' : NULL, ENT_NOQUOTES) ?>
 <?php echo Nette\Templating\Helpers::escapeHtml($hello ? 'yes' : NULL, ENT_NOQUOTES) ?></li>
	<li>Array: <?php echo Nette\Templating\Helpers::escapeHtml($template->join($hello ? array('a', 'b', 'c') : NULL), ENT_NOQUOTES) ?>
 <?php echo Nette\Templating\Helpers::escapeHtml($template->join(array('a', 'b', $hello ? 'c' : NULL)), ENT_NOQUOTES) ?></li>
</ul>

<input /> <input /> <input />

<input checked /> <input checked="checked" />

{ test} {"test} {'test}

<p<?php echo Nette\Utils\Html::el(NULL, array('title' => $hello, 'lang' => isset($lang) ? $lang : NULL))->attributes() ?>> </p>

<p val = <?php if (true): ?>"a"<?php else: ?>"b"<?php endif ?>> </p>

<p val0 val1 = tmp val2=tmp val3="x"></p> </p>

<p val = <?php echo '"' . htmlSpecialChars($xss) . '"' ?>> </p>

<p val =></p><?php echo Nette\Templating\Helpers::escapeHtml($xss, ENT_NOQUOTES) ?></p>

<?php echo Nette\Templating\Helpers::escapeHtml((string) (bool) (float) (int) (array) 10, ENT_NOQUOTES) ?>


<?php if ($hello): ?>
	<?php echo Nette\Templating\Helpers::escapeHtml($hello, ENT_NOQUOTES) ?>

<?php elseif ($any): ?>
	<?php echo Nette\Templating\Helpers::escapeHtml($any, ENT_NOQUOTES) ?>

<?php else: ?>
	none
<?php endif ?>


<?php $iterations = 0; foreach ($iterator = $_l->its[] = new Nette\Iterators\CachingIterator(array(TRUE)) as $foo): $iterations = 0; foreach ($iterator = $_l->its[] = new Nette\Iterators\CachingIterator($people) as $person): ?>
	<?php if ($iterator->isFirst()): ?><ul><?php endif ?>

	<li id="item-<?php echo htmlSpecialChars($iterator->getCounter()) ?>" class="<?php echo htmlSpecialChars($iterator->isOdd() ? 'odd' : 'even') ?>
"><?php echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ?></li>
	<?php if ($iterator->isLast()): ?></ul><?php endif ?>

<?php $iterations++; endforeach; array_pop($_l->its); $iterator = end($_l->its) ;$iterations++; endforeach; array_pop($_l->its); $iterator = end($_l->its) ?>



<style type="text/css">
<!--
#<?php echo Nette\Templating\Helpers::escapeCss($xss) ?> {
	background: blue;
}
-->
</style>



<script>
<!--
alert('</div>');

var prop = <?php echo Nette\Templating\Helpers::escapeJs($people) ?>;

document.getElementById(<?php echo Nette\Templating\Helpers::escapeJs($xss) ?>).style.backgroundColor = 'red';

var html = <?php echo Nette\Templating\Helpers::escapeJs($el) ?>;
-->
</script>


<script>
/* <![CDATA[ */

var prop2 = <?php echo Nette\Templating\Helpers::escapeJs($people) ?>;

/* ]]> */
</script>


<p onclick =
'alert(<?php echo htmlSpecialChars(Nette\Templating\Helpers::escapeJs($xss), ENT_QUOTES) ?>);alert("hello");'
 title='<?php echo htmlSpecialChars($xss, ENT_QUOTES) ?>'
 style =
 "color:<?php echo htmlSpecialChars(Nette\Templating\Helpers::escapeCss($xss)) ?>;"
 rel="<?php echo htmlSpecialChars($xss) ?>"
 onblur="alert(<?php echo htmlSpecialChars(Nette\Templating\Helpers::escapeJs($xss)) ?>)"
 alt='<?php echo htmlSpecialChars($el, ENT_QUOTES) ?>'
 onfocus="alert(<?php echo htmlSpecialChars(Nette\Templating\Helpers::escapeJs($el)) ?>)"
>click on me <?php echo Nette\Templating\Helpers::escapeHtml($xss, ENT_NOQUOTES) ?></p>



<?php $counter = 0 ;if ($_l->extends) { ob_end_clean(); return Nette\Latte\Macros\CoreMacros::includeTemplate($_l->extends, get_defined_vars(), $template)->render(); }
call_user_func(reset($_l->blocks['menu']), $_l, get_defined_vars())  ?>


<?php extract(array('varx' => 2), EXTR_SKIP) ;echo Nette\Templating\Helpers::escapeHtml($varx, ENT_NOQUOTES) ?>

<?php extract(array('varx' => 4), EXTR_SKIP) ;echo Nette\Templating\Helpers::escapeHtml($varx, ENT_NOQUOTES) ?>


<?php $varx = 8 ;echo Nette\Templating\Helpers::escapeHtml($varx, ENT_NOQUOTES) ?>

<?php $varx = 9 ;echo Nette\Templating\Helpers::escapeHtml($varx, ENT_NOQUOTES) ?>


<?php $now = new DateTime ;$foo = $now->format('u') ;extract(array('foo' => $now->format('u')), EXTR_SKIP) ?>



</ul>


<ul title="foreach">
<?php $iterations = 0; foreach ($people as $person): ?>	<li><?php echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ?></li>
<?php $iterations++; endforeach ?>
</ul>

<ul title="for">
<?php for ($i=0;$i<3;$i++): ?>	<li><?php echo Nette\Templating\Helpers::escapeHtml($i, ENT_NOQUOTES) ?></li>
<?php endfor ?>
</ul>

<ul title="white">
<?php while (--$i>0): ?>	<li><?php echo Nette\Templating\Helpers::escapeHtml($i, ENT_NOQUOTES) ?></li>
<?php endwhile ?>
</ul>

<?php if (true): ?><p>
	<div><p>true</div>
</p>
<?php endif ?>

<?php if (true): ?><p>
	<div><p>true</p></div>
</p>
<?php endif ?>

<?php if (false): ?><p>
	<div><p>false</div>
</p>
<?php endif ?>

<?php if (false): ?><p>
	<div><p>false</p></div>
</p>
<?php endif ?>

<?php if (strlen('{$xss}') > 5): ?><p>noLatte</p>
<?php endif ?>

<ul title="if + foreach">
<?php $iterations = 0; foreach ($people as $person): if (strlen($person)===4): ?>
	<li><?php echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ?></li>
<?php endif ;$iterations++; endforeach ?>
</ul>

<ul title="if + inner-if + inner-foreach">
<?php if (empty($iterator)): ?>	<li><?php $iterations = 0; foreach ($people as $person): if (strlen($person)===4): echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ;endif ;$iterations++; endforeach ?></li>
<?php endif ?>
</ul>

<ul title="inner-if + inner-foreach">
<?php $iterations = 0; foreach ($people as $person): if (strlen($person)===4): ?>
	<li><?php echo Nette\Templating\Helpers::escapeHtml($template->lower($person), ENT_NOQUOTES) ?></li>
<?php endif ;$iterations++; endforeach ?></ul>

<?php call_user_func(reset($_l->blocks['bl']), $_l, get_defined_vars())  ?>

<?php if ($_l->ifs[] = (true)): ?><b><?php endif ?>
bold<?php if (array_pop($_l->ifs)): ?></b><?php endif ?>
 <?php if ($_l->ifs[] = (false)): ?><b><?php endif ?>
normal<?php if (array_pop($_l->ifs)): ?></b>
<?php endif ?>

<?php if ($_l->ifs[] = (true)): ?><b<?php if ($_l->tmp = array_filter(array('first'))) echo ' class="' . htmlSpecialChars(implode(" ", array_unique($_l->tmp))) . '"' ?>
><?php endif ?>
bold<?php if (array_pop($_l->ifs)): ?></b>
<?php endif ?>

<?php echo 'test' ?>


{ }

<p<?php if ($_l->tmp = array_filter(array('foo', (false ? 'first' : NULL), 'odd', ($hello ? 'foo' : 'bar')))) echo ' class="' . htmlSpecialChars(implode(" ", array_unique($_l->tmp))) . '"' ?>>n:class</p>

<p<?php if ($_l->tmp = array_filter(array(false ? 'first' : NULL))) echo ' class="' . htmlSpecialChars(implode(" ", array_unique($_l->tmp))) . '"' ?>>n:class empty</p>

<meta />
<?php if (true): ?><meta />
<?php endif ;if ($_l->ifs[] = (true)): ?><meta />
<?php endif ;if (array_pop($_l->ifs)): endif ?>
<meta />
<?php if (true): endif ?><meta />

<?php ob_start() ?>TRUE<?php if (TRUE) ob_end_flush(); else ob_end_clean() ?>

<?php ob_start() ?>FALSE<?php if (FALSE) ob_end_flush(); else ob_end_clean() ?>

<?php ob_start() ?>TRUE<?php ob_start() ?>FALSE<?php if (TRUE) { ob_end_clean(); ob_end_flush(); } else { $_else = ob_get_contents(); ob_end_clean(); ob_end_clean(); echo $_else; } ?>


<?php $iterations = 0; foreach ($iterator = $_l->its[] = new Nette\Iterators\CachingIterator(array(1,2,3)) as $foo): ?>
	<b<?php if ($_l->tmp = array_filter(array($iterator->even ? 'even' : NULL))) echo ' class="' . htmlSpecialChars(implode(" ", array_unique($_l->tmp))) . '"' ?>>item</b>
<?php $iterations++; endforeach; array_pop($_l->its); $iterator = end($_l->its) ?>

<?php $iterations = 0; foreach ($iterator = $_l->its[] = new Nette\Iterators\CachingIterator(array(1, 2, 3)) as $foo): ?>
<p<?php if ($_l->tmp = array_filter(array($iterator->even ? 'even' : NULL))) echo ' class="' . htmlSpecialChars(implode(" ", array_unique($_l->tmp))) . '"' ?>
><?php echo Nette\Templating\Helpers::escapeHtml($foo, ENT_NOQUOTES) ?></p>
<?php $iterations++; endforeach; array_pop($_l->its); $iterator = end($_l->its) ?>

<?php $iterations = 0; foreach (array(0) as $foo): if (1): $iterations = 0; foreach (array(1) as $foo): if ($_l->ifs[] = (0)): ?>
<span><?php endif ;$iterations++; endforeach ;$iterations = 0; foreach (array(2) as $foo): if (2): ?>
Hello<?php endif ;$iterations++; endforeach ;$iterations = 0; foreach (array(1) as $foo): if (array_pop($_l->ifs)): ?></span>
<?php endif ;$iterations++; endforeach ;endif ;$iterations++; endforeach ?>


<ul title="foreach break">
<?php $iterations = 0; foreach ($people as $person): ?>	<li><?php echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ;if (TRUE) { echo "</li>\n"; break; } ?></li>
<?php $iterations++; endforeach ?>
</ul>

<ul title="foreach continue">
<?php $iterations = 0; foreach ($people as $person): ?>	<li><?php echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ;if (TRUE) { echo "</li>\n"; continue; } ?></li>
<?php $iterations++; endforeach ?>
</ul>


<ul title="inner foreach break">
	<li><?php $iterations = 0; foreach ($people as $person): echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ;if (TRUE) break ;$iterations++; endforeach ?></li>
</ul>

<ul title="inner foreach continue">
	<li><?php $iterations = 0; foreach ($people as $person): echo Nette\Templating\Helpers::escapeHtml($person, ENT_NOQUOTES) ;if (TRUE) continue ;$iterations++; endforeach ?></li>
</ul>
