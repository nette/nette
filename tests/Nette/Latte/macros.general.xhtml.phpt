<?php

/**
 * Test: Nette\Latte\Engine: general XHTML test.
 *
 * @author     David Grudl
 * @package    Nette\Latte
 * @keepTrailingSpaces
 */

use Nette\Latte,
	Nette\Templating\FileTemplate,
	Nette\Utils\Html;



require __DIR__ . '/../bootstrap.php';

require __DIR__ . '/Template.inc';



$template = new FileTemplate(__DIR__ . '/templates/general.latte');
$template->registerFilter(new Latte\Engine);
$template->registerHelper('translate', 'strrev');
$template->registerHelper('join', 'implode');
$template->registerHelperLoader('Nette\Templating\Helpers::loader');

$template->hello = '<i>Hello</i>';
$template->xss = 'some&<>"\'/chars';
$template->people = array('John', 'Mary', 'Paul', ']]>');
$template->menu = array('about', array('product1', 'product2'), 'contact');
$template->el = Html::el('div')->title('1/2"');

$path = __DIR__ . '/expected/' . basename(__FILE__, '.phpt');
Assert::match(file_get_contents("$path.phtml"), codefix($template->compile()));
Assert::match(file_get_contents("$path.html"), $template->__toString(TRUE));
