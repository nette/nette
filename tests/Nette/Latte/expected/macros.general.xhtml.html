<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<ul>
	<li>Escaped: &lt;i&gt;Hello&lt;/i&gt;</li>
	<li>Non-escaped: <i>Hello</i></li>
	<li>Escaped expression: &lt;b&gt;hello&lt;/b&gt;</li>
	<li>Non-escaped expression: <b>hello</b></li>
	<li>Array access: Mary</li>
	<li>Html: <div title='1/2"'></div></li>
	<li>Translated: yraM</li>
	<li>Translated string: joha</li>
	<li>Non-escaped and translated: yraM</li>
	<li>Condition: yes yes</li>
	<li>Array: abc abc</li>
</ul>

<input /> <input /> <input />

<input checked /> <input checked="checked" />

{ test} {"test} {'test}

<p title="<i>Hello</i>"> </p>

<p val = "a"> </p>

<p val0 val1 = tmp val2=tmp val3="x"></p> </p>

<p val = "some&amp;&lt;&gt;&quot;'/chars"> </p>

<p val =></p>some&amp;&lt;&gt;"'/chars</p>

1

	&lt;i&gt;Hello&lt;/i&gt;


	<ul>
	<li id="item-1" class="odd">John</li>


	<li id="item-2" class="even">Mary</li>


	<li id="item-3" class="odd">Paul</li>


	<li id="item-4" class="even">]]&gt;</li>
	</ul>



<style type="text/css">
<!--
#some\&\<\>\"\'\/chars {
	background: blue;
}
-->
</style>



<script>
<!--
alert('</div>');

var prop = ["John","Mary","Paul","]]\x3E"];

document.getElementById("some&<>\"'\/chars").style.backgroundColor = 'red';

var html = "<div title='1\/2\"'><\/div>";
-->
</script>


<script>
/* <![CDATA[ */

var prop2 = ["John","Mary","Paul","]]\x3E"];

/* ]]> */
</script>


<p onclick =
'alert(&quot;some&amp;&lt;&gt;\&quot;&#039;\/chars&quot;);alert("hello");'
 title='some&amp;&lt;&gt;&quot;&#039;/chars'
 style =
 "color:some\&amp;\&lt;\&gt;\&quot;\'\/chars;"
 rel="some&amp;&lt;&gt;&quot;'/chars"
 onblur="alert(&quot;some&amp;&lt;&gt;\&quot;'\/chars&quot;)"
 alt='&lt;div title=&#039;1/2&quot;&#039;&gt;&lt;/div&gt;'
 onfocus="alert(&quot;&lt;div title='1\/2\&quot;'&gt;&lt;\/div&gt;&quot;)"
>click on me some&amp;&lt;&gt;"'/chars</p>



<ul>
	<li>0 about</li>
	<li>1  <ul>
	<li>2 product1</li>
	<li>3 product2</li>
</ul>
 </li>
	<li>2 contact</li>
</ul>


2
2

8
9




</ul>


<ul title="foreach">
	<li>John</li>
	<li>Mary</li>
	<li>Paul</li>
	<li>]]&gt;</li>
</ul>

<ul title="for">
	<li>0</li>
	<li>1</li>
	<li>2</li>
</ul>

<ul title="white">
	<li>2</li>
	<li>1</li>
</ul>

<p>
	<div><p>true</div>
</p>

<p>
	<div><p>true</p></div>
</p>



<p>noLatte</p>

<ul title="if + foreach">
	<li>John</li>
	<li>Mary</li>
	<li>Paul</li>
</ul>

<ul title="if + inner-if + inner-foreach">
	<li>JohnMaryPaul</li>
</ul>

<ul title="inner-if + inner-foreach">
	<li>john</li>
	<li>mary</li>
	<li>paul</li>
</ul>

<ul title="block + if + foreach">
	<li>John</li>
	<li>Mary</li>
	<li>Paul</li>
</ul>

<b>bold</b> normal
<b class="first">bold</b>

test

{ }

<p class="foo odd">n:class</p>

<p>n:class empty</p>

<meta />
<meta />
<meta />
<meta />
<meta />

TRUE

TRUE

	<b>item</b>
	<b class="even">item</b>
	<b>item</b>

<p>1</p>
<p class="even">2</p>
<p>3</p>

Hello

<ul title="foreach break">
	<li>John</li>
</ul>

<ul title="foreach continue">
	<li>John</li>
	<li>Mary</li>
	<li>Paul</li>
	<li>]]&gt;</li>
</ul>


<ul title="inner foreach break">
	<li>John</li>
</ul>

<ul title="inner foreach continue">
	<li>JohnMaryPaul]]&gt;</li>
</ul>
